assets/src/block-editor/setupBlockEditorValidation.js

```diff
@@ -0,0 +1,182 @@
+const {registerPlugin} = wp.plugins;
+const {PluginPrePublishPanel} = wp.editor;
+const {select, dispatch, subscribe, useSelect} = wp.data;
+const {__} = wp.i18n;
+
+// Object to store block-specific validations
+const blockValidations = {};
+
+// Messages array to hold validation messages
+let messages = [];
+let canPublish = true;
+
+// Post types that require a featured image
+const POST_TYPES_WITH_REQUIRED_FEATURED_IMAGE = ['p4_action', 'post', 'page', 'campaign'];
+
+/**
+ * Setup the block editor validation.
+ * It subscribes to changes in the editor and performs validation checks on title, featured image, and blocks.
+ */
+export const setupBlockEditorValidation = () => {
+  subscribe(() => {
+    const {getEditedPostAttribute, getCurrentPostType, getEditedPostContent} = select('core/editor');
+    const {getBlocks} = select('core/block-editor');
+
+    // Get the necessary data from the editor state
+    const title = getEditedPostAttribute('title');
+    const featuredImage = getEditedPostAttribute('featured_media');
+    const postType = getCurrentPostType();
+    const postContent = getEditedPostContent();
+    const blocks = getBlocks();
+    const currentMessages = [];
+
+    // Validation for title (must not be empty)
+    const invalidTitle = !title || title.trim().length <= 0;
+    if (invalidTitle) {
+      currentMessages.push('Title is required.');
+    }
+
+    // Check if the post requires a featured image and if it's missing
+    const hasImageInContent = /<img.+wp-image-(\d+).*>/i.test(postContent);
+    const needsFeaturedImage = POST_TYPES_WITH_REQUIRED_FEATURED_IMAGE.includes(postType) &&
+      !featuredImage &&
+      !hasImageInContent;
+
+    if (needsFeaturedImage) {
+      currentMessages.push('Featured image is required.');
+    }
+
+    // Get an array of Topic Link blocks without background image
+    const topicLinkBlocksNeedsImage = blocks
+      .filter(block => block.name === 'planet4-blocks/topic-link')
```

Comment:
What do you think if you filter only once?
```suggestion
      .filter(block => block.name === 'planet4-blocks/topic-link' && !block.attributes.imageUrl)
```

Reply:
Change applied!